File: src/main/java/com/demonwav/mcdev/platform/AbstractModuleType.java
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2017 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.platform;

import com.demonwav.mcdev.insight.generation.ui.EventGenerationPanel;

import com.intellij.codeInspection.ex.EntryPointsManager;
import com.intellij.codeInspection.ex.EntryPointsManagerBase;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.JDOMExternalizableStringList;
import com.intellij.psi.PsiClass;
import org.apache.commons.lang.builder.ToStringBuilder;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import java.awt.Color;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import javax.swing.Icon;

public abstract class AbstractModuleType<T extends AbstractModule> {

    @NotNull
    private final String groupId;
    @NotNull
    private final String artifactId;
    @NotNull
    protected final LinkedHashMap<String, Color> colorMap = new LinkedHashMap<>();

    public AbstractModuleType(@NotNull final String groupId, @NotNull final String artifactId) {
        this.groupId = groupId;
        this.artifactId = artifactId;
    }

    @NotNull
    @Contract(pure = true)
    public String getGroupId() {
        return groupId;
    }

    @NotNull
    @Contract(pure = true)
    public String getArtifactId() {
        return artifactId;
    }

    @Contract(pure = true)
    public abstract PlatformType getPlatformType();

    @Contract(pure = true)
    public abstract Icon getIcon();

    @Contract(pure = true)
    public boolean hasIcon() {
        return true;
    }

    @Contract(pure = true)
    public abstract String getId();

    @NotNull
    @Contract(pure = true)
    public abstract List<String> getIgnoredAnnotations();

    @NotNull
    @Contract(pure = true)
    public abstract List<String> getListenerAnnotations();

    @NotNull
    @Contract(pure = true)
    public Map<String, Color> getClassToColorMappings() {
        return this.colorMap;
    }

    @NotNull
    public abstract T generateModule(@NotNull Module module);

    public void performCreationSettingSetup(@NotNull Project project) {
        final JDOMExternalizableStringList annotations = ((EntryPointsManagerBase)EntryPointsManager.getInstance(project)).ADDITIONAL_ANNOTATIONS;
        getIgnoredAnnotations().stream().filter(annotation -> !annotations.contains(annotation)).forEach(annotations::add);
    }

    @NotNull
    @Contract(pure = true)
    public EventGenerationPanel getEventGenerationPanel(@NotNull PsiClass chosenClass) {
        return new EventGenerationPanel(chosenClass);
    }

    @Contract(pure = true)
    public boolean isEventGenAvailable() {
        return false;
    }

    @NotNull
    @Contract(pure = true)
    public String getDefaultListenerName(@NotNull PsiClass psiClass) {
        //noinspection ConstantConditions
        return "on" + psiClass.getName().replace("Event", "");
    }

    @Override
    public String toString() {
        return new ToStringBuilder(this)
            .append("groupId", groupId)
            .append("artifactId", artifactId)
            .toString();
    }
}

File: src/main/kotlin/com/demonwav/mcdev/platform/AbstractModuleType.kt
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2019 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.platform

import com.demonwav.mcdev.facet.MinecraftFacet
import com.demonwav.mcdev.insight.generation.ui.EventGenerationPanel
import com.demonwav.mcdev.util.findContainingClass
import com.intellij.codeInspection.ex.EntryPointsManager
import com.intellij.codeInspection.ex.EntryPointsManagerBase
import com.intellij.openapi.module.ModuleUtilCore
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiClass
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import java.awt.Color
import java.util.LinkedHashMap
import javax.swing.Icon
import org.apache.commons.lang.builder.ToStringBuilder
import org.jetbrains.annotations.Contract

abstract class AbstractModuleType<out T : AbstractModule>(val groupId: String, val artifactId: String) {

    val colorMap = LinkedHashMap<String, Color>()

    abstract val platformType: PlatformType

    abstract val icon: Icon?

    open val hasIcon = true

    abstract val id: String

    abstract val ignoredAnnotations: List<String>

    abstract val listenerAnnotations: List<String>

    val classToColorMappings: Map<String, Color>
        get() = this.colorMap

    abstract fun generateModule(facet: MinecraftFacet): T

    fun performCreationSettingSetup(project: Project) {
        if (project.isDisposed) {
            return
        }
        val manager = EntryPointsManager.getInstance(project)
        val annotations = (manager as? EntryPointsManagerBase)?.ADDITIONAL_ANNOTATIONS as? MutableList<String> ?: return
        ignoredAnnotations.asSequence()
            .filter { annotation -> !annotations.contains(annotation) }
            .forEach { annotations.add(it) }
    }

    open fun getEventGenerationPanel(chosenClass: PsiClass): EventGenerationPanel {
        return EventGenerationPanel(chosenClass)
    }

    @get:Contract(pure = true)
    open val isEventGenAvailable: Boolean
        get() = false

    open fun getDefaultListenerName(psiClass: PsiClass) = "on" + psiClass.name!!.replace("Event", "")

    override fun toString(): String {
        return ToStringBuilder(this)
            .append("groupId", groupId)
            .append("artifactId", artifactId)
            .toString()
    }

    /**
     * Given any PsiElement, determine if it resides in a module of this [AbstractModuleType].

     * @param element The element to check.
     * *
     * @return True if this element resides in a module of this type
     */
    fun isInModule(element: PsiElement): Boolean {
        val module = ModuleUtilCore.findModuleForPsiElement(element) ?: return false

        val facet = MinecraftFacet.getInstance(module)
        return facet != null && facet.isOfType(this)
    }

    protected fun defaultNameForSubClassEvents(psiClass: PsiClass): String {
        val isInnerClass = psiClass.parent !is PsiFile

        val name = StringBuilder()
        if (isInnerClass) {
            val containingClass = psiClass.parent.findContainingClass()
            if (containingClass != null && containingClass.name != null) {
                name.append(containingClass.name!!.replace("Event", ""))
            }
        }

        var className = psiClass.name!!
        if (className.startsWith(name.toString())) {
            className = className.substring(name.length)
        }
        name.append(className.replace("Event", ""))

        name.insert(0, "on")
        return name.toString()
    }
}
